pytorch_lightning/callbacks/pruning.py

```diff
@@ -0,0 +1,396 @@
+        ELSE, pruning_fn will be resolved into its function counterpart from `torch.nn.utils.prune`.
+
+        """
+        if self.use_global_unstructured:
+            pruning_fn = _PYTORCH_PRUNING_METHOD[pruning_fn]
+            self._global_kwargs = kwargs
+            return pruning_fn
+        else:
+            return ModelPruning._wrap_pruning_fn(_PYTORCH_PRUNING_FUNCTIONS[pruning_fn], **kwargs)
+
+    @staticmethod
+    def _wrap_pruning_fn(pruning_fn, *args, **kwargs):
+        return partial(pruning_fn, **kwargs)
+
+    def _make_pruning_permanent(self):
+        for module, param_name in self.parameters_to_prune:
+            pytorch_prune.remove(module, param_name)
+
+    def _resolve_amount(self, current_epoch: int) -> float:
+        if isinstance(self.amount, Callable):
+            amount_fn = self.amount
+            amount = amount_fn(current_epoch)
+        else:
+            amount = self.amount
+        return amount
+
+    def _restore_original_weights(self, module: nn.Module, orig_module: nn.Module, tensor_name: str):
+        """
+        "The lottery ticket hypothesis" (https://arxiv.org/pdf/1803.03635.pdf) algorithm:
+
+            1. Randomly initialize a neural network f(x;θ0)(where θ0 ∼Dθ).
+            2. Train the network for j iterations, arriving at parameters θj .
+            3. Prune p% of the parameters in θj , creating a mask m.
+            4. Reset the remaining parameters to their values in θ0, creating the winning ticket f(x; m⊙θ0).
+
+        This function is responsible of step 4.
+        """
+        trained = getattr(module, tensor_name)
+        orig = getattr(orig_module, tensor_name)
+        trained.data = orig.data.to(trained.device)
+
+    def apply_lottery_ticket_hypothesis(self):
+        for (module, tensor_name), (orig_module, _) in zip(self.parameters_to_prune, self._parameters_to_prune):
+            self._restore_original_weights(module, orig_module, tensor_name)
+
+    def _apply_local_pruning(self, amount: float):
+        for module, param in self.parameters_to_prune:
+            self.pruning_fn(module, name=param, amount=amount)
+
+    def _resolve_global_kwargs(self, amount: float):
+        kwargs = {}
+        self._global_kwargs["amount"] = amount
+        params = inspect.signature(self.pruning_fn).parameters
+        for p_name in params:
+            if p_name != "self":
+                param = self._global_kwargs.get(p_name)
+                if param is not None:
+                    kwargs[p_name] = param
+        return kwargs
+
+    def _apply_global_pruning(self, amount: float):
+        pytorch_prune.global_unstructured(
+            self.parameters_to_prune,
+            pruning_method=self.pruning_fn,
+            **self._resolve_global_kwargs(amount)
+        )
+
+    def apply_pruning(self, trainer: 'pl.Trainer', pl_module: LightningModule):
+        amount = self._resolve_amount(trainer.current_epoch)
```

Comment:
I think it might be useful to skip pruning if amount is 0. This would also allow controlling the pruning frequency

Reply:
Done !